File: tools/automation/src/safe10-evidence.ts
import { createHash } from "node:crypto";
import { chmod, mkdir, readFile, readdir, rename, stat, unlink, writeFile } from "node:fs/promises";
import path from "node:path";

import {
  internalCommandSpec,
  type AutomationCategory,
  type Safe10Evidence,
} from "./contracts.generated.js";
import { portFromDetectorOutput } from "./detector.js";
import type { ProcessPort } from "./process.js";
import {
  safe10AttemptProductionDigest,
  safe10CurrentInventory,
  safe10ProductionFragments,
} from "./safe10-source-inventory.js";
import { assertWithinWorkspace } from "./workspace.js";

type JsonObject = Readonly<Record<string, unknown>>;
type FailureCategory = Extract<AutomationCategory, "evidence_invalid" | "process_failed" | "timeout">;

const expectedPlan = "docs/parity/work-plans/20260818T122819Z-SAFE-10/PLAN.md";
const expectedPlanSha256 = "ca4230f4668843be0d1a433b061e6dddaf9fb25b3d318094e30945ca71648690";
const expectedTask = "task-parity-safe10-prerequisite-readiness";
const expectedAttemptRoot = "scratch/stat003-scoreboard/attempt-003";
const expectedDetectorOutput = "scratch/stat003-scoreboard/wrapper-003/detector.stdout";
const expectedAttemptPlan = "docs/parity/work-plans/20260818T102038Z-STAT-003/PLAN.md";
const expectedAttemptPlanSha256 = "41ca445088dcf15c4c1c46e504a754c61260e7575eb16ccf68e0edb0fc742879";
const expectedAttemptClosure = "docs/parity/work-plans/20260818T102038Z-STAT-003/CLOSURE.md";
const expectedAttemptSourceCommit = "60a56d4935ced15eeb5ec6950b1ad4ea35fdf223";
const expectedReferenceCommit = "c1915b0a63bfabebdb95a515cedfee05146c1d50";
const expectedProjection = "docs/parity/evidence/safe10-prerequisite-readiness/safe10-projection.json";

export type Safe10EvidenceOptions = Readonly<{
  attemptRoot: string;
  detectorOutput: string;
  attemptPlan: string;
  attemptClosure: string;
  projection: string;
}>;

export class Safe10EvidenceError extends Error {
  public constructor(
    public readonly category: FailureCategory,
    message: string,
    public readonly publicValue: Readonly<Record<string, unknown>> = {
      stage: "safe10_projection",
      projection_published: false,
    },
  ) {
    super(message);
    this.name = "Safe10EvidenceError";
  }
}

function failure(category: FailureCategory, message: string): Safe10EvidenceError {
  return new Safe10EvidenceError(category, message);
}

function sha256(value: string | Buffer): string {
  return createHash("sha256").update(value).digest("hex");
}

function object(value: unknown, context: string): JsonObject {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw failure("evidence_invalid", `${context} must be an object`);
  }
  return value as JsonObject;
}

function requiredString(value: JsonObject, field: string, context: string): string {
  const candidate = value[field];
  if (typeof candidate !== "string" || candidate.length === 0) {
    throw failure("evidence_invalid", `${context} string field is invalid`);
  }
  return candidate;
}

function requiredInteger(value: JsonObject, field: string, context: string): number {
  const candidate = value[field];
  if (typeof candidate !== "number" || !Number.isSafeInteger(candidate) || candidate < 0) {
    throw failure("evidence_invalid", `${context} integer field is invalid`);
  }
  return candidate;
}

function requiredBoolean(value: JsonObject, field: string, context: string): boolean {
  const candidate = value[field];
  if (typeof candidate !== "boolean") {
    throw failure("evidence_invalid", `${context} boolean field is invalid`);
  }
  return candidate;
}

async function readJson(candidate: string, context: string): Promise<{
  readonly document: string;
  readonly value: JsonObject;
}> {
  const document = await readFile(candidate, "utf8");
  try {
    return { document, value: object(JSON.parse(document), context) };
  } catch (error) {
    if (error instanceof Safe10EvidenceError) throw error;
    throw failure("evidence_invalid", `${context} is malformed`);
  }
}

async function requireMode(candidate: string, mode: number, directory: boolean): Promise<void> {
  const metadata = await stat(candidate);
  if ((directory ? !metadata.isDirectory() : !metadata.isFile())
    || (metadata.mode & 0o777) !== mode) {
    throw failure("evidence_invalid", "SAFE-10 protected evidence mode is invalid");
  }
}

async function requirePrivateTreeModes(root: string): Promise<void> {
  await requireMode(root, 0o700, true);
  for (const entry of await readdir(root, { withFileTypes: true })) {
    const candidate = path.join(root, entry.name);
    if (entry.isDirectory()) await requirePrivateTreeModes(candidate);
    else if (entry.isFile()) await requireMode(candidate, 0o600, false);
    else throw failure("evidence_invalid", "SAFE-10 protected evidence contains a special file");
  }
}

async function childText(
  processPort: ProcessPort,
  program: string,
  args: readonly string[],
  context: string,
): Promise<string> {
  let outcome;
  try {
    outcome = await processPort.run(internalCommandSpec(program, [...args], (value) => value));
  } catch {
    throw failure("process_failed", `${context} launch failed`);
  }
  if (outcome.timedOut) throw failure("timeout", `${context} timed out`);
  if (outcome.exitCode !== 0) throw failure("evidence_invalid", `${context} failed`);
  return outcome.stdout;
}

function exactFreshness(value: JsonObject, context: string): {
  readonly power: boolean;
  readonly voltage: boolean;
  readonly current: boolean;
  readonly chip: boolean;
  readonly vr: boolean;
  readonly fan: boolean;
} {
  return {
    power: requiredBoolean(value, "power_watts", context),
    voltage: requiredBoolean(value, "bus_voltage_volts", context),
    current: requiredBoolean(value, "current_amps", context),
    chip: requiredBoolean(value, "chip_temp_celsius", context),
    vr: requiredBoolean(value, "vr_temp_celsius", context),
    fan: requiredBoolean(value, "fan_rpm", context),
  };
}

export async function projectSafe10Evidence(
  workspaceRoot: string,
  options: Safe10EvidenceOptions,
  processPort: ProcessPort,
  gitProgram: string,
  validatorProgram: string,
): Promise<Safe10Evidence> {
  const attemptRoot = assertWithinWorkspace(workspaceRoot, options.attemptRoot);
  const detectorOutput = assertWithinWorkspace(workspaceRoot, options.detectorOutput);
  const attemptPlan = assertWithinWorkspace(workspaceRoot, options.attemptPlan);
  const attemptClosure = assertWithinWorkspace(workspaceRoot, options.attemptClosure);
  const projection = assertWithinWorkspace(workspaceRoot, options.projection);
  const candidate = `${projection}.candidate`;
  if (path.relative(workspaceRoot, attemptRoot) !== expectedAttemptRoot
    || path.relative(workspaceRoot, detectorOutput) !== expectedDetectorOutput
    || path.relative(workspaceRoot, attemptPlan) !== expectedAttemptPlan
    || path.relative(workspaceRoot, attemptClosure) !== expectedAttemptClosure
    || path.relative(workspaceRoot, projection) !== expectedProjection) {
    throw failure("evidence_invalid", "SAFE-10 immutable path binding is invalid");
  }
  try {
    await stat(projection);
    throw failure("evidence_invalid", "SAFE-10 projection already exists");
  } catch (error) {
    if (error instanceof Safe10EvidenceError) throw error;
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
  await requirePrivateTreeModes(attemptRoot);
  await requireMode(detectorOutput, 0o600, false);
  await portFromDetectorOutput(workspaceRoot, options.detectorOutput);

  const [taskDocument, planDocument, attemptPlanDocument, attemptClosureDocument] =
    await Promise.all([
      readFile(path.join(workspaceRoot, "TASKS.md"), "utf8"),
      readFile(path.join(workspaceRoot, expectedPlan), "utf8"),
      readFile(attemptPlan, "utf8"),
      readFile(attemptClosure, "utf8"),
    ]);
  const heading = `### ${expectedTask} |`;
  const taskStart = taskDocument.indexOf(heading);
  const taskEnd = taskDocument.indexOf("\n### ", taskStart + heading.length);
  const taskBlock = taskDocument.slice(taskStart, taskEnd === -1 ? taskDocument.length : taskEnd);
  if (taskStart === -1 || taskDocument.indexOf(heading, taskStart + heading.length) !== -1
    || !taskBlock.includes(expectedPlan)
    || sha256(planDocument) !== expectedPlanSha256
    || sha256(attemptPlanDocument) !== expectedAttemptPlanSha256
    || !attemptClosureDocument.includes(`- Plan SHA-256: \`${expectedAttemptPlanSha256}\``)
    || !attemptClosureDocument.includes(expectedAttemptSourceCommit)) {
    throw failure("evidence_invalid", "SAFE-10 task or plan lineage is invalid");
  }

  const campaignRoot = path.join(attemptRoot, "campaign");
  const [result, network, observations, diagnostics] = await Promise.all([
    readJson(path.join(campaignRoot, "campaign-result.json"), "campaign result"),
    readJson(path.join(campaignRoot, "campaign-network.private.json"), "campaign network"),
    readJson(path.join(campaignRoot, "campaign-observations.private.json"), "campaign observations"),
    readJson(path.join(campaignRoot, "campaign-diagnostics.private.json"), "campaign diagnostics"),
  ]);
  const seal = (await readFile(path.join(campaignRoot, "campaign-result.sha256"), "utf8")).trim();
  if (seal !== sha256(result.document)
    || requiredString(result.value, "network_continuity_sha256", "campaign result") !== sha256(network.document)
    || requiredString(result.value, "observations_sha256", "campaign result") !== sha256(observations.document)
    || requiredString(result.value, "diagnostics_sha256", "campaign result") !== sha256(diagnostics.document)) {
    throw failure("evidence_invalid", "SAFE-10 campaign seal or digest chain is invalid");
  }

  const freshness = exactFreshness(
    object(result.value["observation_freshness"], "observation freshness"),
    "observation freshness",
  );
  const requirements = exactFreshness(
    object(result.value["observation_requirements"], "observation requirements"),
    "observation requirements",
  );
  const terminal = object(observations.value["terminal_marker"], "terminal marker");
  const readiness = object(terminal["readiness_transition"], "readiness transition");
  const qualifiedCount = requiredInteger(result.value, "qualified_candidate_count", "campaign result");
  const prerequisites = {
    power_watts_required: requirements.power,
    bus_voltage_required: requirements.voltage,
    current_required: requirements.current,
    chip_temperature_required: requirements.chip,
    vr_temperature_required: requirements.vr,
    fan_rpm_required: requirements.fan,
    power_watts_fresh: freshness.power,
    bus_voltage_fresh: freshness.voltage,
    current_fresh: freshness.current,
    chip_temperature_fresh: freshness.chip,
    vr_temperature_fresh: freshness.vr,
    fan_rpm_fresh: freshness.fan,
    fresh_observation_count: requiredInteger(result.value, "fresh_observation_count", "campaign result"),
    safety_fresh: requiredString(result.value, "safety", "campaign result") === "fresh",
    readiness_unblocked: requiredString(readiness, "current_blocker", "readiness transition") === "none",
    session_running_primary: requiredString(readiness, "session_phase", "readiness transition") === "running_primary",
    hardware_ready: requiredString(readiness, "hardware_state", "readiness transition") === "ready",
    readiness_safety_fresh: requiredString(readiness, "safety_sample", "readiness transition") === "fresh",
    observation_epoch_advanced: requiredString(readiness, "observation_epoch", "readiness transition") === "advanced",
    pending_observation_recovered: requiredBoolean(readiness, "pending_observation_recovered", "readiness transition"),
    active_ms: requiredInteger(result.value, "active_ms", "campaign result"),
    required_window_count: requiredInteger(network.value, "required_window_count", "campaign network"),
    covered_window_count: requiredInteger(network.value, "covered_window_count", "campaign network"),
    work_renewal_valid: requiredBoolean(network.value, "work_renewal_valid", "campaign network"),
    active_state_valid: requiredBoolean(network.value, "active_state_valid", "campaign network"),
    network_safety_valid: requiredBoolean(network.value, "safety_valid", "campaign network"),
    watchdog_valid: requiredBoolean(network.value, "watchdog_valid", "campaign network"),
    qualified_candidate_observed: qualifiedCount > 0,
    accepted_submit_observed: requiredString(result.value, "submit_outcome", "campaign result") === "accepted",
    terminal_http_valid: requiredBoolean(network.value, "terminal_http_valid", "campaign network"),
    terminal_websocket_valid: requiredBoolean(network.value, "terminal_websocket_valid", "campaign network"),
    terminal_pool_persisted: requiredBoolean(network.value, "terminal_pool_persisted", "campaign network"),
    final_terminal_consumed: requiredBoolean(network.value, "final_terminal_consumed", "campaign network"),
    serial_finished_observed: requiredBoolean(network.value, "serial_finished_observed", "campaign network"),
  };
  if (requiredString(result.value, "schema", "campaign result") !== "mining-campaign-result-v16"
    || requiredString(result.value, "status", "campaign result") !== "accepted"
    || requiredString(result.value, "stage", "campaign result") !== "live-share"
    || requiredString(result.value, "profile", "campaign result") !== "conservative"
    || requiredInteger(result.value, "duration_seconds", "campaign result") !== 600
    || requiredString(result.value, "runtime_identity", "campaign result") !== "trusted"
    || requiredString(result.value, "network_status", "campaign result") !== "accepted"
    || requiredString(result.value, "safe_stop", "campaign result") !== "confirmed"
    || requiredString(result.value, "usb_cleanup", "campaign result") !== "ready"
    || requiredString(network.value, "status", "campaign network") !== "accepted"
    || requiredString(network.value, "correlation_failure", "campaign network") !== "none"
    || requiredString(network.value, "watchdog_failure", "campaign network") !== "none"
    || requiredString(diagnostics.value, "runtime_attestation_mixed_reset_reason", "campaign diagnostics") !== "none"
    || requiredString(diagnostics.value, "panic_signature", "campaign diagnostics") !== "none"
    || requiredInteger(diagnostics.value, "panic_signature_count", "campaign diagnostics") !== 0) {
    throw failure("evidence_invalid", "SAFE-10 live campaign boundary is incomplete");
  }

  const currentCommit = (await childText(processPort, gitProgram, ["rev-parse", "HEAD"], "current source identity")).trim();
  const pushedCommit = (await childText(processPort, gitProgram, ["rev-parse", "origin/main"], "pushed source identity")).trim();
  const referenceCommit = (await childText(
    processPort,
    gitProgram,
    ["-C", path.join(workspaceRoot, "reference/esp-miner"), "rev-parse", "HEAD"],
    "reference identity",
  )).trim();
  const dirty = (await childText(
    processPort,
    gitProgram,
    ["status", "--porcelain", "--untracked-files=no"],
    "source cleanliness",
  )).trim();
  if (currentCommit !== pushedCommit || referenceCommit !== expectedReferenceCommit || dirty !== "") {
    throw failure("evidence_invalid", "SAFE-10 clean pushed source identity is invalid");
  }
  const currentInventory = await safe10CurrentInventory(workspaceRoot);
  const attemptDocuments = new Map<string, Buffer>();
  for (const relative of safe10ProductionFragments.keys()) {
    const document = await childText(
      processPort,
      gitProgram,
      ["show", `${expectedAttemptSourceCommit}:${relative}`],
      "attempt source inventory",
    );
    attemptDocuments.set(relative, Buffer.from(document));
  }
  const attemptSourceDigest = safe10AttemptProductionDigest(attemptDocuments);
  const sourceCompatible = attemptSourceDigest === currentInventory.productionDigest;
  const evidence: Safe10Evidence = {
    schema_version: "bitaxe-safe10-evidence-v1",
    board: 205,
    attempt_ordinal: 3,
    attempt_source_commit: expectedAttemptSourceCommit,
    current_source_commit: currentCommit,
    reference_commit: referenceCommit,
    workflow: {
      schema_version: "bitaxe-workflow-identity-v1",
      command: "project-safe10-evidence",
      request_sha256: sha256(JSON.stringify({
        plan: expectedPlanSha256,
        attempt_plan: expectedAttemptPlanSha256,
        attempt_closure: sha256(attemptClosureDocument),
        campaign_result: sha256(result.document),
        current_source: currentInventory.digest,
        attempt_source: attemptSourceDigest,
      })),
    },
    source: {
      plan_sha256: expectedPlanSha256,
      attempt_plan_sha256: expectedAttemptPlanSha256,
      attempt_closure_sha256: sha256(attemptClosureDocument),
      campaign_result_sha256: sha256(result.document),
      campaign_network_sha256: sha256(network.document),
      campaign_observations_sha256: sha256(observations.document),
      campaign_diagnostics_sha256: sha256(diagnostics.document),
      current_source_inventory_sha256: currentInventory.digest,
      attempt_source_inventory_sha256: attemptSourceDigest,
      source_semantics_current: true,
      reference_semantics_current: true,
      attempt_source_compatible: sourceCompatible,
      source_path_count: currentInventory.pathCount,
      production_path_count: safe10ProductionFragments.size,
      reference_path_count: 2,
    },
    prerequisites,
    detector_admitted: true,
    runtime_identity: "trusted",
    campaign_stage: "live-share",
    campaign_profile: "conservative",
    campaign_status: "accepted",
    network_status: "accepted",
    safe_stop_confirmed: true,
    cleanup_complete: true,
    hardware_rerun_used: false,
    protected_modes_valid: true,
    redaction_status: "passed",
  };
  await mkdir(path.dirname(projection), { recursive: true });
  try {
    await writeFile(candidate, `${JSON.stringify(evidence, null, 2)}\n`, {
      encoding: "utf8",
      mode: 0o600,
      flag: "wx",
    });
    await childText(processPort, validatorProgram, [candidate], "SAFE-10 evidence validator");
    await chmod(candidate, 0o644);
    await rename(candidate, projection);
    return evidence;
  } catch (error) {
    try {
      await unlink(candidate);
    } catch (cleanupError) {
      if ((cleanupError as NodeJS.ErrnoException).code !== "ENOENT") throw cleanupError;
    }
    if (error instanceof Safe10EvidenceError) throw error;
    throw failure("evidence_invalid", "SAFE-10 evidence publication failed");
  }
}

File: crates/bitaxe-automation-contracts/src/safe10_evidence.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{AutomationCommand, WorkflowIdentity, SAFE10_EVIDENCE_SCHEMA};

#[derive(Debug, Clone, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
pub struct Safe10SourceEvidence {
    pub plan_sha256: String,
    pub attempt_plan_sha256: String,
    pub attempt_closure_sha256: String,
    pub campaign_result_sha256: String,
    pub campaign_network_sha256: String,
    pub campaign_observations_sha256: String,
    pub campaign_diagnostics_sha256: String,
    pub current_source_inventory_sha256: String,
    pub attempt_source_inventory_sha256: String,
    pub source_semantics_current: bool,
    pub reference_semantics_current: bool,
    pub attempt_source_compatible: bool,
    pub source_path_count: u8,
    pub production_path_count: u8,
    pub reference_path_count: u8,
}

#[derive(Debug, Clone, Copy, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
pub struct Safe10PrerequisiteEvidence {
    pub power_watts_required: bool,
    pub bus_voltage_required: bool,
    pub current_required: bool,
    pub chip_temperature_required: bool,
    pub vr_temperature_required: bool,
    pub fan_rpm_required: bool,
    pub power_watts_fresh: bool,
    pub bus_voltage_fresh: bool,
    pub current_fresh: bool,
    pub chip_temperature_fresh: bool,
    pub vr_temperature_fresh: bool,
    pub fan_rpm_fresh: bool,
    pub fresh_observation_count: u8,
    pub safety_fresh: bool,
    pub readiness_unblocked: bool,
    pub session_running_primary: bool,
    pub hardware_ready: bool,
    pub readiness_safety_fresh: bool,
    pub observation_epoch_advanced: bool,
    pub pending_observation_recovered: bool,
    pub active_ms: u64,
    pub required_window_count: u8,
    pub covered_window_count: u8,
    pub work_renewal_valid: bool,
    pub active_state_valid: bool,
    pub network_safety_valid: bool,
    pub watchdog_valid: bool,
    pub qualified_candidate_observed: bool,
    pub accepted_submit_observed: bool,
    pub terminal_http_valid: bool,
    pub terminal_websocket_valid: bool,
    pub terminal_pool_persisted: bool,
    pub final_terminal_consumed: bool,
    pub serial_finished_observed: bool,
}

impl Safe10PrerequisiteEvidence {
    fn complete(self) -> bool {
        self.power_watts_required
            && self.bus_voltage_required
            && self.current_required
            && self.chip_temperature_required
            && !self.vr_temperature_required
            && self.fan_rpm_required
            && self.power_watts_fresh
            && self.bus_voltage_fresh
            && self.current_fresh
            && self.chip_temperature_fresh
            && !self.vr_temperature_fresh
            && self.fan_rpm_fresh
            && self.fresh_observation_count == 5
            && self.safety_fresh
            && self.readiness_unblocked
            && self.session_running_primary
            && self.hardware_ready
            && self.readiness_safety_fresh
            && self.observation_epoch_advanced
            && !self.pending_observation_recovered
            && self.active_ms >= 600_000
            && self.required_window_count == 20
            && self.covered_window_count == 20
            && self.work_renewal_valid
            && self.active_state_valid
            && self.network_safety_valid
            && self.watchdog_valid
            && self.qualified_candidate_observed
            && self.accepted_submit_observed
            && self.terminal_http_valid
            && self.terminal_websocket_valid
            && self.terminal_pool_persisted
            && self.final_terminal_consumed
            && self.serial_finished_observed
    }
}

#[derive(Debug, Clone, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct Safe10Evidence {
    pub schema_version: String,
    pub board: u16,
    pub attempt_ordinal: u8,
    pub attempt_source_commit: String,
    pub current_source_commit: String,
    pub reference_commit: String,
    pub workflow: WorkflowIdentity,
    pub source: Safe10SourceEvidence,
    pub prerequisites: Safe10PrerequisiteEvidence,
    pub detector_admitted: bool,
    pub runtime_identity: String,
    pub campaign_stage: String,
    pub campaign_profile: String,
    pub campaign_status: String,
    pub network_status: String,
    pub safe_stop_confirmed: bool,
    pub cleanup_complete: bool,
    pub hardware_rerun_used: bool,
    pub protected_modes_valid: bool,
    pub redaction_status: String,
}

impl Safe10Evidence {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.schema_version != SAFE10_EVIDENCE_SCHEMA
            || self.board != 205
            || self.attempt_ordinal != 3
            || self.workflow.command != AutomationCommand::ProjectSafe10Evidence
        {
            return Err("SAFE-10 identity is invalid");
        }
        for commit in [
            &self.attempt_source_commit,
            &self.current_source_commit,
            &self.reference_commit,
        ] {
            if !is_lower_hex(commit, 40) {
                return Err("SAFE-10 commit identity is invalid");
            }
        }
        for digest in [
            &self.workflow.request_sha256,
            &self.source.plan_sha256,
            &self.source.attempt_plan_sha256,
            &self.source.attempt_closure_sha256,
            &self.source.campaign_result_sha256,
            &self.source.campaign_network_sha256,
            &self.source.campaign_observations_sha256,
            &self.source.campaign_diagnostics_sha256,
            &self.source.current_source_inventory_sha256,
            &self.source.attempt_source_inventory_sha256,
        ] {
            if !is_lower_hex(digest, 64) {
                return Err("SAFE-10 digest identity is invalid");
            }
        }
        if !self.source.source_semantics_current
            || !self.source.reference_semantics_current
            || !self.source.attempt_source_compatible
            || self.source.source_path_count != 19
            || self.source.production_path_count != 9
            || self.source.reference_path_count != 2
        {
            return Err("SAFE-10 source evidence is incomplete");
        }
        if !self.prerequisites.complete() {
            return Err("SAFE-10 prerequisite quorum is incomplete");
        }
        if !self.detector_admitted
            || self.runtime_identity != "trusted"
            || self.campaign_stage != "live-share"
            || self.campaign_profile != "conservative"
            || self.campaign_status != "accepted"
            || self.network_status != "accepted"
            || !self.safe_stop_confirmed
            || !self.cleanup_complete
            || self.hardware_rerun_used
            || !self.protected_modes_valid
            || self.redaction_status != "passed"
        {
            return Err("SAFE-10 live evidence is incomplete");
        }
        Ok(())
    }
}

fn is_lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn evidence() -> Safe10Evidence {
        Safe10Evidence {
            schema_version: SAFE10_EVIDENCE_SCHEMA.to_owned(),
            board: 205,
            attempt_ordinal: 3,
            attempt_source_commit: "a".repeat(40),
            current_source_commit: "b".repeat(40),
            reference_commit: "c".repeat(40),
            workflow: WorkflowIdentity {
                schema_version: "bitaxe-workflow-identity-v1".to_owned(),
                command: AutomationCommand::ProjectSafe10Evidence,
                request_sha256: "d".repeat(64),
            },
            source: Safe10SourceEvidence {
                plan_sha256: "e".repeat(64),
                attempt_plan_sha256: "f".repeat(64),
                attempt_closure_sha256: "1".repeat(64),
                campaign_result_sha256: "2".repeat(64),
                campaign_network_sha256: "3".repeat(64),
                campaign_observations_sha256: "4".repeat(64),
                campaign_diagnostics_sha256: "5".repeat(64),
                current_source_inventory_sha256: "6".repeat(64),
                attempt_source_inventory_sha256: "7".repeat(64),
                source_semantics_current: true,
                reference_semantics_current: true,
                attempt_source_compatible: true,
                source_path_count: 19,
                production_path_count: 9,
                reference_path_count: 2,
            },
            prerequisites: Safe10PrerequisiteEvidence {
                power_watts_required: true,
                bus_voltage_required: true,
                current_required: true,
                chip_temperature_required: true,
                vr_temperature_required: false,
                fan_rpm_required: true,
                power_watts_fresh: true,
                bus_voltage_fresh: true,
                current_fresh: true,
                chip_temperature_fresh: true,
                vr_temperature_fresh: false,
                fan_rpm_fresh: true,
                fresh_observation_count: 5,
                safety_fresh: true,
                readiness_unblocked: true,
                session_running_primary: true,
                hardware_ready: true,
                readiness_safety_fresh: true,
                observation_epoch_advanced: true,
                pending_observation_recovered: false,
                active_ms: 600_000,
                required_window_count: 20,
                covered_window_count: 20,
                work_renewal_valid: true,
                active_state_valid: true,
                network_safety_valid: true,
                watchdog_valid: true,
                qualified_candidate_observed: true,
                accepted_submit_observed: true,
                terminal_http_valid: true,
                terminal_websocket_valid: true,
                terminal_pool_persisted: true,
                final_terminal_consumed: true,
                serial_finished_observed: true,
            },
            detector_admitted: true,
            runtime_identity: "trusted".to_owned(),
            campaign_stage: "live-share".to_owned(),
            campaign_profile: "conservative".to_owned(),
            campaign_status: "accepted".to_owned(),
            network_status: "accepted".to_owned(),
            safe_stop_confirmed: true,
            cleanup_complete: true,
            hardware_rerun_used: false,
            protected_modes_valid: true,
            redaction_status: "passed".to_owned(),
        }
    }

    #[test]
    fn complete_live_prerequisite_evidence_passes() {
        // Arrange
        let evidence = evidence();

        // Act / Assert
        assert_eq!(evidence.validate(), Ok(()));
    }

    #[test]
    fn every_required_prerequisite_is_fail_closed() {
        for mutate in [
            |value: &mut Safe10PrerequisiteEvidence| value.power_watts_fresh = false,
            |value: &mut Safe10PrerequisiteEvidence| value.bus_voltage_fresh = false,
            |value: &mut Safe10PrerequisiteEvidence| value.current_fresh = false,
            |value: &mut Safe10PrerequisiteEvidence| value.chip_temperature_fresh = false,
            |value: &mut Safe10PrerequisiteEvidence| value.fan_rpm_fresh = false,
            |value: &mut Safe10PrerequisiteEvidence| value.readiness_unblocked = false,
            |value: &mut Safe10PrerequisiteEvidence| value.hardware_ready = false,
            |value: &mut Safe10PrerequisiteEvidence| value.work_renewal_valid = false,
        ] {
            // Arrange
            let mut evidence = evidence();
            mutate(&mut evidence.prerequisites);

            // Act / Assert
            assert_eq!(
                evidence.validate(),
                Err("SAFE-10 prerequisite quorum is incomplete")
            );
        }
    }

    #[test]
    fn source_drift_and_missing_protected_modes_fail() {
        // Arrange
        let mut source_drift = evidence();
        source_drift.source.attempt_source_compatible = false;
        let mut bad_modes = evidence();
        bad_modes.protected_modes_valid = false;

        // Act / Assert
        assert_eq!(
            source_drift.validate(),
            Err("SAFE-10 source evidence is incomplete")
        );
        assert_eq!(
            bad_modes.validate(),
            Err("SAFE-10 live evidence is incomplete")
        );
    }
}
